backend/database.js

```diff
@@ -37,38 +39,53 @@ if (!fs.existsSync(dataFolder)) {
   fs.mkdirSync(dataFolder, { mode: 0o750 })
 }
 
+// Streams stored contract log entries since the given entry hash (inclusive!).
 sbp('sbp/selectors/register', {
-  'backend/db/streamEntriesAfter': async function (contractID: string, hash: string): Promise<*> {
+  'backend/db/streamEntriesAfter': async function (contractID: string, hash: string, requestedLimit: number = MAX_EVENTS_AFTER): Promise<*> {
+    const limit = Math.min(requestedLimit, process.env.MAX_EVENTS_BATCH_SIZE ?? Number.POSITIVE_INFINITY)
     const latestHEADinfo = await sbp('chelonia/db/latestHEADinfo', contractID)
     if (!latestHEADinfo) {
       throw Boom.notFound(`contractID ${contractID} doesn't exist!`)
     }
-    let { HEAD: currentHEAD } = latestHEADinfo
+    // Number of entries pushed.
+    let counter = 0
+    let currentHash = hash
     let prefix = '['
     // NOTE: if this ever stops working you can also try Readable.from():
     // https://nodejs.org/api/stream.html#stream_stream_readable_from_iterable_options
-    return new Readable({
-      async read (): any {
-        try {
-          const entry = await sbp('chelonia/db/getEntry', currentHEAD)
-          const json = `"${strToB64(entry.serialize())}"`
-          if (currentHEAD !== hash) {
-            this.push(prefix + json)
-            currentHEAD = entry.head().previousHEAD
-            prefix = ','
-          } else {
-            this.push(prefix + json + ']')
+    const stream = new Readable({
+      read (): void {
+        if (currentHash && counter < limit) {
+          sbp('chelonia/db/getEntry', currentHash).then(async entry => {
+            if (entry) {
+              this.push(`${prefix}"${strToB64(entry.serialize())}"`)
+              prefix = ','
+              counter++
+              currentHash = await sbp('chelonia/db/get', `next=${currentHash}`)
+            } else {
+              this.push(counter > 0 ? ']' : '[]')
+              this.push(null)
+            }
+          }).catch(e => {
+            console.error(`[backend] streamEntriesAfter: read(): ${e.message}:`, e)
+            this.push(counter > 0 ? ']' : '[]')
             this.push(null)
-          }
-        } catch (e) {
-          console.error(e, `read(): ${e.message}`)
-          this.push(']')
+          })
+        } else {
+          this.push(counter > 0 ? ']' : '[]')
           this.push(null)
         }
       }
     })
+    // $FlowFixMe[prop-missing]
+    stream.headers = {
+      'shelter-headinfo-head': latestHEADinfo.HEAD,
+      'shelter-headinfo-height': latestHEADinfo.height
+    }
+    return stream
   },
-  'backend/db/streamEntriesBefore': async function (before: string, limit: number): Promise<*> {
+  'backend/db/streamEntriesBefore': async function (before: string, requestedLimit: number): Promise<*> {
+    let limit = Math.min(requestedLimit, process.env.MAX_EVENTS_BATCH_SIZE ?? Number.POSITIVE_INFINITY)
```

Comment:
We need to set `MAX_EVENTS_BATCH_SIZE` as a constant in the code, let's set it to 500